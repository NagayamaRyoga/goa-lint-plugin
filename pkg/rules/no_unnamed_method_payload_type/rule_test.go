package no_unnamed_method_payload_type_test

import (
	"log"
	"os"
	"testing"

	"github.com/NagayamaRyoga/goalint/pkg/rules/no_unnamed_method_payload_type"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"goa.design/goa/v3/dsl"
	"goa.design/goa/v3/eval"
	"goa.design/goa/v3/expr"
)

func TestRule(t *testing.T) {
	t.Parallel()

	var (
		AddNumbersPayload = dsl.Type("AddNumbersPayload", func() {
			dsl.Attribute("a", expr.Int)
			dsl.Attribute("b", expr.Int)
			dsl.Required("a", "b")
		})

		methodWithNamedPayload = dsl.Service("calc", func() {
			dsl.Method("add_numbers", func() {
				dsl.Payload(AddNumbersPayload)
			})
		})

		methodWithUnnamedPayload = dsl.Service("calc2", func() {
			dsl.Method("add_numbers2", func() {
				dsl.Payload(func() {
					dsl.Attribute("a", expr.Int)
					dsl.Attribute("b", expr.Int)
					dsl.Required("a", "b")
				})
			})
		})

		methodWithIntPayload = dsl.Service("calc3", func() {
			dsl.Method("add_numbers3", func() {
				dsl.Payload(dsl.Int)
			})
		})
	)

	// given
	err := eval.RunDSL()
	require.NoError(t, err)

	testCases := []struct {
		description string
		service     *expr.ServiceExpr
		wantReports int
	}{
		{
			description: "success",
			service:     methodWithNamedPayload,
			wantReports: 0,
		},
		{
			description: "failed/object",
			service:     methodWithUnnamedPayload,
			wantReports: 1,
		},
		{
			description: "failed/int",
			service:     methodWithIntPayload,
			wantReports: 1,
		},
	}
	for _, tc := range testCases {
		t.Run(tc.description, func(t *testing.T) {
			t.Parallel()

			snaps := snaps.WithConfig(snaps.Dir("testdata"))

			logger := log.New(os.Stdout, "", 0)
			cfg := no_unnamed_method_payload_type.NewConfig()
			rule := no_unnamed_method_payload_type.NewRule(logger, cfg)

			// when
			assert.Len(t, tc.service.Methods, 1)

			got := rule.WalkMethodExpr(tc.service.Methods[0])
			assert.Len(t, got, tc.wantReports)
			snaps.MatchSnapshot(t, got.String())
		})
	}
}
