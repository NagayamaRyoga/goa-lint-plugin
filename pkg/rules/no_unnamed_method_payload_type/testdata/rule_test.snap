
[TestRule/success - 1]

---

[TestRule/failed - 1]
Found 1 errors and 0 warnings
[NoUnnamedMethodPayloadType]: error in service "calc2" method "add_numbers2": Method payload should be an user defined type

---

[TestRule/failed/int - 1]
Found 1 errors and 0 warnings
[NoUnnamedMethodPayloadType]: error in service "calc3" method "add_numbers3": Method payload should be an user defined type

---

[TestRule/failed/object - 1]
Found 1 errors and 0 warnings
[NoUnnamedMethodPayloadType]: error in service "calc2" method "add_numbers2": Method payload should be an user defined type

---
